File: rewrite-python/src/test/java/org/openrewrite/python/ChangeMethodNameTest.java
/*
 * Copyright 2024 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.python;

import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.openrewrite.DocumentExample;
import org.openrewrite.test.RecipeSpec;
import org.openrewrite.test.RewriteTest;

import static org.openrewrite.python.Assertions.python;

class ChangeMethodNameTest implements RewriteTest {

    @Override
    public void defaults(RecipeSpec spec) {
        spec.recipe(new ChangeMethodName("print", "println", true));
    }

    @Disabled("Flaky")
    @DocumentExample
    @SuppressWarnings("PyUnresolvedReferences")
    @Test
    void renameMethod() {
        rewriteRun(
          spec -> spec.parser(PythonParser.builder().logCompilationWarningsAndErrors(true)),
          python(
            """
            class Foo:
                def foo(self) :
                    print("hello")
            """,
            """
            class Foo:
                def foo(self) :
                    println("hello")
            """
          )
        );
    }
}

File: rewrite-java/src/test/kotlin/org/openrewrite/java/ChangeMethodNameTest.kt
/*
 * Copyright 2020 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java

import org.junit.jupiter.api.Test

open class ChangeMethodNameTest : JavaParser() {

    private val b: String = """
                class B {
                   public void singleArg(String s) {}
                   public void arrArg(String[] s) {}
                   public void varargArg(String... s) {}
                }
            """.trimIndent()

    @Test
    fun refactorMethodNameForMethodWithSingleArgDeclarative() {
        val a = """
            class A {
               public void test() {
                   new B().singleArg("boo");
               }
            }
        """.trimIndent()

        val cu = parse(a, b)

        val fixed = cu.refactor()
                .visit(ChangeMethodName().apply { setMethod("B singleArg(String)"); setName("bar") })
                .fix().fixed

        assertRefactored(fixed, """
            class A {
               public void test() {
                   new B().bar("boo");
               }
            }
        """)
    }

    @Test
    fun refactorMethodNameForMethodWithSingleArg() {
        val a = """
            class A {
               public void test() {
                   new B().singleArg("boo");
               }
            }
        """.trimIndent()

        val cu = parse(a, b)

        val fixed = cu.refactor()
                .fold(cu.findMethodCalls("B singleArg(String)")) { ChangeMethodName.Scoped(it, "bar") }
                .fix().fixed

        assertRefactored(fixed, """
            class A {
               public void test() {
                   new B().bar("boo");
               }
            }
        """)
    }

    @Test
    fun refactorMethodNameForMethodWithArrayArg() {
        val a = """
            class A {
               public void test() {
                   new B().arrArg(new String[] {"boo"});
               }
            }
        """.trimIndent()

        val cu = parse(a, b)

        val fixed = cu.refactor()
                .fold(cu.findMethodCalls("B arrArg(String[])")) { ChangeMethodName.Scoped(it, "bar") }
                .fix().fixed

        assertRefactored(fixed, """
            class A {
               public void test() {
                   new B().bar(new String[] {"boo"});
               }
            }
        """)
    }

    @Test
    fun refactorMethodNameForMethodWithVarargArg() {
        val a = """
            class A {
               public void test() {
                   new B().varargArg("boo", "again");
               }
            }
        """.trimIndent()

        val cu = parse(a, b)

        val fixed = cu.refactor()
                .fold(cu.findMethodCalls("B varargArg(String...)")) { ChangeMethodName.Scoped(it, "bar") }
                .fix().fixed

        assertRefactored(fixed, """
            class A {
               public void test() {
                   new B().bar("boo", "again");
               }
            }
        """)
    }

    @Test
    fun refactorMethodNameWhenMatchingAgainstMethodWithNameThatIsAnAspectjToken() {
        val b = """
            class B {
               public void error() {}
               public void foo() {}
            }
        """.trimIndent()

        val a = """
            class A {
               public void test() {
                   new B().error();
               }
            }
        """.trimIndent()

        val cu = parse(a, b)
        val fixed = cu.refactor()
                .fold(cu.findMethodCalls("B error()")) { ChangeMethodName.Scoped(it, "foo") }
                .fix().fixed

        assertRefactored(fixed, """
            class A {
               public void test() {
                   new B().foo();
               }
            }
        """)
    }
}
